File: zig/15-01/main.zig
// 0.15.2
const std = @import("std");

fn getChange(symbol: u8) i8 {
    if (symbol == '(') return 1;
    if (symbol == ')') return -1;
    return 0;
}

fn partOne(line: []const u8) i16 {
    var floor: i16 = 0;

    for (line) |symbol| {
        floor = floor + getChange(symbol);
        //std.debug.print("{u}\n", .{symbol});
    }

    return floor;
}

fn partTwo(line: []const u8) usize {
    var floor: i16 = 0;

    for (line, 0..) |symbol, i| {
        floor = floor + getChange(symbol);
        if (floor == -1) {
            return i + 1;
        }
    }

    return 0;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    var file = try std.fs.cwd().openFile("input.txt", .{ .mode = .read_only });
    defer file.close();

    const stat = try file.stat();
    const buff = try file.readToEndAlloc(allocator, stat.size);
    defer allocator.free(buff);

    var lines = std.mem.splitSequence(u8, buff, "\n");
    const line: []const u8 = lines.next() orelse return;

    std.debug.print("partOne: {d}\n", .{partOne(line)});
    std.debug.print("partTwo: {d}\n", .{partTwo(line)});
}

File: haskell/15-01/Main.hs
import Text.Printf

floorChange :: Int -> Char -> Int
floorChange floor '(' = floor + 1
floorChange floor ')' = floor - 1
floorChange floor _ = floor

partOne :: [Char] -> Int -> Int
partOne xs floor = foldl floorChange floor xs

partTwo :: [Char] -> Int -> Int -> Int
partTwo _ floor pos | floor == -1 = pos
partTwo (x : xs) floor pos = partTwo xs (floorChange floor x) (pos + 1)

main :: IO ()
main = do
  input <- readFile "input.txt"
  printf "PartOne: %d\n" $ partOne input 0
  printf "PartTwo: %d\n" $ partTwo input 0 0
